### PR TITLE
auth: First draft of krb/gssapi  authentication for Ceph

### PR DESCRIPTION
**Pull request by oliveiradan**
We rely on a keytab file and name resolution for the KDC.

### CODE REVIEW

**Comment by liewegas** [comment]
This is great!

- needs rebase
- it looks like the authorizer bits aren't implemented yet.  makes sense; this is step 1
- there's a fair bit of dead code in there.  at some point we'll want to clean it all out.  it's probably okay for now since most of those sections are to-be-implemented.
- would it be possible to include some doc describing how to set up an environment to test/exercise some of this code?

**Comment by oliveiradan** [comment]
@liewegas , yes. I talked to Joao this morning about it (rebasing, clean ups, docs on setup and QA). I will keep this thread with all updates.

File: src/auth/ldap_krb/KrbClientHandler.cpp [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+      << "KrbClientHandler::build_request() " << dendl; 
+
+  RWLock::RLocker lck(lock);
+  KrbTokenBlob krb_token; 
+  KrbRequest krb_request; 
+
+  krb_request.m_request_type = 
+      static_cast<int>(gss_client_auth::
+                          GSSAuthenticationRequest::GSS_TOKEN);
+  ::encode(krb_request, buff_list);
```
Comment (adamemerson) [inline src/auth/ldap_krb/KrbClientHandler.cpp]: This is deprecated and probably doesn't work. The encode/decode methods have either been moved into the ceph namespace or the namespace of the types they reference.

Either use do using ceph::encode/using ceph::decode and let Argument Dependent Lookup do the rest, or fully qualify the call for whatever type you're using.

File: src/auth/ldap_krb/KrbClientHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+                                           gss_minor_status));
+      ldout(cct, SUBSYSTEM_ZERO) 
+          << "ERROR: KrbClientHandler::handle_response() "
+             "[gss_import_name(gss_service_name)] failed! " 
+          << gss_major_status << " " 
+          << gss_minor_status << " " 
+          << status_str 
+          << dendl;
+    }
+  }
```
Comment (adamemerson) [inline src/auth/ldap_krb/KrbClientHandler.cpp]: Minor style nit, but everywhere else in the code we do

`} else {`

**Pull request closed without merging**